### PR TITLE
Learn from null coalescing to bool constants

### LINKED ISSUE #37394: Nullable Check didn't check for ?? false
**Version Used**: 16.2.0-preview-19278-01+d635043bd

**Steps to Reproduce**:
```
mkdir NullableTest && cd NullableTest
dotnet new console
```
Then enable NullableContextOptions, and write code:
```csharp
using System;

namespace NullableTest
{
    class InnerClass
    {
        public bool TestField;
    }
    class Test
    {
        public InnerClass? InnerField;
    }
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
        }

        static bool Test()
        {
            var x = new Test();
            if (x.InnerField?.TestField ?? false) // obviously if x.InnerField is null, it won't enter the `if` statement so there's no nullable-related issues
            {
                return x.InnerField.TestField;
            }
            return false;
        }
    }
}
```

```
dotnet build
```

**Expected Behavior**:
No warnings and no errors.

**Actual Behavior**:
```
Program.cs(25,24): warning CS8602: Dereference of a possibly null reference.
```

**Additional Information**
The if (xx?.yy ?? false) is commonly used in C#, I hope that Roslyn can support checking nullability for this syntax

### PR DESCRIPTION
**Pull request by RikkiGibson**
Resolves #37394 

Put intuitively: if the RHS of a `??` operator is a constant boolean `b`, we know that the value `!b` could only be produced by the LHS of the `??` operator being not-null.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -33000,6 +33000,73 @@ void Test(C? x, C? y)
...
+        [WorkItem(26624, "https://github.com/dotnet/roslyn/issues/26624")]
+        public void ConditionalBranching_20()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Test1(CL1? x1)
+    {
+        _ = x1?.Next?.Next?.P1 ?? false
+            ? x1.Next.Next.ToString()
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >x1.Next.Next [](start = 14, length = 12)

Could you test `x1.ToString()` on both branches? Maybe `? (x.Next.Next.ToString() + x.ToString()) ...`

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Done

**Comment by CyrusNajmabadi** [comment]
Question about this stuff.  Is this being driven through the LDM at all?  Are we collecting all the rules so that we can doc them and explain tehm to people picking up nullable and trying to figure out which coding patterns convey nullability information and which do not?  

Thanks!

**Comment by RikkiGibson** [comment]
@CyrusNajmabadi this particular change has not gone through LDM. Other changes probably have, especially those that have deeper-running design implications, e.g. if you conditionally access off of a not-null receiver, its flow state can change to maybe-null afterwards.

It would be good to catalogue all the minor behaviors that are more or less just scattered around NullableWalker and various GitHub issues. AFAIK we have not done this. But the nullable analysis work is expected to be more ad-hoc and less rigorous than something like definite assignment analysis.

**Comment by CyrusNajmabadi** [comment]
>AFAIK we have not done this. But the nullable analysis work is expected to be more ad-hoc and less rigorous than something like definite assignment analysis.

Interesting.  I both like the idea that we can be flexible here and more open to changes in this area without full LDM approval.  However, we should def at least try to doc the rules in a reasonable fashion.  After all, we will have to have something people can read to actually understand how this works.  I would expect MSDN articles and references and whatnot :)  Those should ideally be as accurate as we can make things for any given release :)

Anyways, this all sounds good to me.  I would just actually recommend having a tracking doc for this sort of thing so it doesn't get lost in the churn :)

**Pull request merged**